Fix #1840: Introduce a nicer model for configuring network presentation. (#1848)
kroxylicious/kroxylicious: docs/modules/configuring/con-configuring-vc-listeners.adoc

[id='con-configuring-vc-listeners-{context}']
= Configuring virtual cluster listeners

[role="_abstract"]

Clients make connections to virtual cluster listeners. Each listener defines a bootstrap address.  This is the address
that the client makes its initial connection to. The listener also provides a means for the clients to talk to the
proxied brokers. There is a choice about how this is done:

* *Port Identifies Node* - the listener binds separate ports - one for each broker as well as a separate one for bootstrap.
  Clients make connections to the different _port numbers_ as they need to converse with each broker.
* *SNI Host Identifies Node* - the listener assigns a distinct hostname to each broker.
  Clients make connections to the different _hostnames_ as they need to converse with each broker.

>   The listener uses SNI (Server Name Identification) to work out which broker the client wishes to talk to

Reviewer: I like the change in terminology to `X Identifies Node` a lot clearer than anything we have had previously.

>   The listener uses SNI (Server Name Identification) to work out which broker the client wishes to talk to

Reviewer 2: SNI=Server Name Indication?

IMPORTANT: You must make sure that the listener's bootstrap address and generated broker addresses are resolvable
           and routable by the Kafka Client.  You must also make sure firewall rules permit traffic to required port numbers.

> == Port Identifies Node

Reviewer: I think we should note somewhere in this section that there is an implicit mapping between port and broker id and maybe that sparse ID ranges cause problems.

Author: if you are using ranges,we  don't actually have that problem, do we?

io.kroxylicious.proxy.internal.clusternetworkaddressconfigprovider.RangeAwarePortPerNodeClusterNetworkAddressConfigProviderTest#discoveryAddressMapMultipleRanges

Reviewer: We sort of still do, in that we need a hint to people to define multiple ranges if they have disparate id groups or huge id numbers we need to know the base offset. 

Though at that point it might be fine in the details of the configuration.

that each listener has its own range of port numbers and these do not overlap with the range used by another listener, or
the listener of another virtual cluster.

[id='con-configuring-vc-listeners-port-identifies-node-{context}']
.Example port identifies node configuration

[source,yaml]
----
# ...
virtualClusters:
  my-cluster-proxy:
    # ...
    listener:
    - name: mylistener
      portIdentifiesNode:
        bootstrapAddress: localhost:9192 # <1>
    # ...
# ...
----
<1> The bootstrap address used by Kafka clients.

With the example configuration above, the listener exposes a target cluster of up to three brokers with
node ids `0`, `1`, `2`.   The advertised address is defaulted to that of the bootstrap host name.  Port
numbers are assigned sequentially beginning at bootstrap port number + 1.

The listener exposes the following three broker addresses:

|===
|Node Id|Broker Address

|0
|localhost:9193

|1
|localhost:9194

|2
|localhost:9195
|===

.Example port identifies node configuration with customized advertised broker address and port

[source,yaml]
----
# ...
virtualClusters:
  my-cluster-proxy:
    # ...
    listener:
    - name: mylistener
      portIdentifiesNode:

>         bootstrapAddress: localhost:9192 # <1>

Reviewer: nit: I think `localhost` is a poor choice for an example

>         advertisedBrokerAddressPattern: mycluster.kafka.com # <2>

Reviewer: IIRC `example.com` is reserved for exactly this sort of usage. While its not actively used `kafka.com` is a registered domain...

        - name: mixed
          start: 1
          end: 3
        - name: brokers
          start: 100
          end: 103

    # ...
# ...
----
<1> The bootstrap address used by Kafka clients.
<2> (Optional) The advertised broker address pattern used to form broker addresses. If not defined, it defaults to the hostname part of the bootstrap address.
<3> (Optional) One or more node id ranges.  If omitted, it defaults to a single range of node IDs 0..2 (inclusive).

With the example configuration above, the listener exposes a target cluster of up to six nodes with
node ids `1..3` and `101..103`.   The advertised broker address is defined as `mycluster.example.com`.  Port
numbers are assigned sequentially beginning at `9193` (bootstrap port number + 1)

The listener exposes the following six broker addresses:

|===
|Node Id|Broker Address

|1
|mycluster.example.com:9193

|2
|mycluster.example.com:9194

|3
|mycluster.example.com:9195

|101
|mycluster.example.com:9196

|102
|mycluster.example.com:9197

|103
|mycluster.example.com:9198

|===

Finally, the `advertisedBrokerAddressPattern` configuration parameter accepts the `$(nodeId)` replacement token, which is optional.

> If included, `$(nodeId)` is replaced by the broker's https://kafka.apache.org/documentation/#brokerconfigs_node.id[`node.id`] (or https://kafka.apache.org/documentation/#brokerconfigs_broker.id[`broker.id`]) in the target cluster.

Author: In general probably not, but I think in this case it is important that the reader understands how the addresses are formed.

[id='con-configuring-vc-listeners-snihost-identifies-node-{context}']

== Server Name Identification (SNI) Host Identifies Node

In the SNI Host Identifies Node scheme, unique broker host names are used to know where to route the traffic.
As this scheme relies on SNI, which is a TLS extension, TLS connections are required. It cannot be used with plain text connections.

In this scheme, you may share the port across many virtual cluster listeners, or you can define a separate port
per virtual cluster listener.  You cannot use a port which is used by a virtual cluster listener using
the Port Identifies Node scheme.

> IMPORTANT: When using this scheme, you have the responsibility to make sure that DNS for bootstrap and brokers resolve
> to an IP address that is routed to the Proxy. Wildcard DNS is one way to achieve this.

Reviewer 2: ```suggestion
IMPORTANT: When using this scheme, you have the responsibility to make sure that DNS for bootstrap and brokers resolve
to an IP address that is routed to the proxy. Wildcard DNS is one way to achieve this.
```

.Example SNI Host Identifies Node configuration
[source,yaml]
----
# ...
virtualClusters:
  my-cluster-proxy:
    # ...
    listeners:
    - name: mylistener
      sniHostIdentifiesNode:
        bootstrapAddress: mycluster.example.com:9192 # <1>
        advertisedBrokerAddressPattern: mybroker-$(nodeId).mycluster.example.com <2>
      tls:
         key: ... <3>
    # ...
# ...
----
<1> The bootstrap address used by Kafka clients.
<2> The advertised broker address pattern used to form broker addresses. It must include the placeholder $(nodeId) which
    will be substituted for the node id.
<3> TLS configuration.

With the example configuration above, the listener accepts all traffic on port 9192. Any TLS connections arriving with
the SNI of `mycluster.example.com` will be routed as boostrap.  Any connections arriving with SNI matching
`mybroker-$(nodeId).mycluster.example.com` will be routed to the upstream broker with the same node id.
It exposes a target cluster with any number of brokers.  It does not need prior knowledge of the node IDs used by
the brokers.

> The listener exposes the following broker addresses like this:

Reviewer 2: ```suggestion
The listener exposes the following broker addresses:
```

|===
|Node Id|Broker Address

|0
|mybroker-0.mycluster.example.com:9192

|...
|...

|_n_
|mybroker-_n_.mycluster.example.com:9192

|===

> .Example SNI routing address provider configuration with customized advertised port

Reviewer 2: ```suggestion
.Example SNI Host Identifies Node configuration with customized advertised port
```

[source,yaml]
----
# ...
virtualClusters:
  my-cluster-proxy:
    # ...
    listeners:
    - name: mylistener
      sniHostIdentifiesNode:
        bootstrapAddress: mycluster.example.com:9192 # <1>
        advertisedBrokerAddressPattern: mybroker-$(nodeId).mycluster.example.com:443 <2>
      tls:
         key: ... <3>
    # ...
# ...
----
<1> The bootstrap address used by Kafka clients.
<2> The advertised broker address pattern and port number used to form broker addresses. It must include the placeholder $(node)
    which will be substituted for the node id.
<3> TLS configuration.

> This configuration instructs Kroxylicious to listen on port 9192, but advertise brokers of this virtual cluster as

Reviewer 2: ```suggestion
With the example configuration above, Kroxylicious is instructed to listen on port 9192, but advertise brokers of this virtual cluster as
```

> being available on port 443.  A use-case for this feature is where a network intermediary (such as another proxy or

Reviewer 2: ```suggestion
being available on port 443. This feature is useful where a network intermediary (such as another proxy or
```

load balancer) is port forwarding.

> The listener exposes the following broker addresses like this:

Reviewer 2: ```suggestion
The listener exposes the following broker addresses:
```
